protocol/src/main/java/org/geysermc/mcprotocollib/network/net/NetClientSession.java

```diff
@@ -0,0 +1,153 @@
+
+import java.net.SocketAddress;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+
+public class NetClientSession extends NetSession implements ClientSession {
+    private static EventLoopGroup EVENT_LOOP_GROUP;
+
+    /**
+     * See {@link EventLoopGroup#shutdownGracefully(long, long, TimeUnit)}
+     */
+    private static final int SHUTDOWN_QUIET_PERIOD_MS = 100;
+    private static final int SHUTDOWN_TIMEOUT_MS = 500;
+
+    protected final SocketAddress bindAddress;
+    protected final ProxyInfo proxy;
+    protected final PacketCodecHelper codecHelper;
+
+    public NetClientSession(SocketAddress remoteAddress, PacketProtocol protocol) {
+        this(remoteAddress, null, protocol);
+    }
+
+    public NetClientSession(SocketAddress remoteAddress, PacketProtocol protocol, ProxyInfo proxy) {
+        this(remoteAddress, null, protocol, proxy);
+    }
+
+    public NetClientSession(SocketAddress remoteAddress, SocketAddress bindAddress, PacketProtocol protocol) {
+        this(remoteAddress, bindAddress, protocol, null);
+    }
+
+    public NetClientSession(SocketAddress remoteAddress, SocketAddress bindAddress, PacketProtocol protocol, ProxyInfo proxy) {
+        this(remoteAddress, bindAddress, protocol, proxy, DefaultPacketHandlerExecutor.createExecutor());
+    }
+
+    public NetClientSession(SocketAddress remoteAddress, SocketAddress bindAddress, PacketProtocol protocol, ProxyInfo proxy, Executor packetHandlerExecutor) {
+        super(remoteAddress, protocol, packetHandlerExecutor);
+        this.bindAddress = bindAddress;
+        this.proxy = proxy;
+        this.codecHelper = protocol.createHelper();
+    }
+
+    @Override
+    public void connect(boolean wait) {
+        if (this.disconnected) {
+            throw new IllegalStateException("Session has already been disconnected.");
+        }
+
+        final Bootstrap bootstrap = new Bootstrap()
+            .channelFactory(getChannelFactory())
+            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, getFlag(BuiltinFlags.CLIENT_CONNECT_TIMEOUT, 30) * 1000)
+            .group(getEventLoopGroup())
+            .remoteAddress(remoteAddress)
+            .localAddress(bindAddress)
+            .handler(getChannelHandler());
+
+        setOptions(bootstrap);
+
+        CompletableFuture<Void> handleFuture = new CompletableFuture<>();
+        bootstrap.connect().addListener((futureListener) -> {
+            if (!futureListener.isSuccess()) {
+                exceptionCaught(null, futureListener.cause());
+            }
+
+            handleFuture.complete(null);
+        });
+
+        if (wait) {
+            handleFuture.join();
+        }
+    }
+
+    @Override
+    public PacketCodecHelper getCodecHelper() {
+        return this.codecHelper;
+    }
+
+    protected EventLoopGroup getEventLoopGroup() {
+        createEventLoopGroup();
```

Comment:
any particular reason the event group null check was moved to the method instead of being here?

Reply:
Better isolates EVENT_LOOP_GROUP to only be referenced from one method. Also imo makes this method more readable for devs looking to override this.